### DISCUSSION TITLE
Accessing Request in handler as well as `FromRequest`

### DISCUSSION THREAD

**Discussion by somehowchris** [tokio-rs/axum#930] [Q&A] [answered]
I have come across a weird behavior so I’m curious if it’s by design and if so why?

## Context
I’m writing a library to make it easy to write Cloud Foundry Route Services which are basically an app on cf acting as a guard and reverse proxy. [Theres already a crate](https://crates.io/crates/hyper-reverse-proxy) basically just proxying a `http::Request`

## Behavior seen

To call my proxying function
```rust
pub async fn proxy_request( req: Request<Body> ) { 
```
I need to have the request so:
```rust
pub async fn handle(req: Request<_>) {
```

Until now, everything works ok. If I would like to have a header at my disposal
```rust 
async fn handle(req: Request<_>, TypedHeader(meta): TypedHeader<ProxyMetadata>) {
```

The request failes with the message:
```log
Header of type `x-cf-proxy-metadata` was missin
```

Now If I remove the `Request` from the handler it works just fine.

> Full example at https://github.com/somehowchris/cf-route-service/blob/main/examples/counter.rs

## Reason why this sucks

Now I get it `req.headers().get(xyz)` would to basically the job, but its not the style of axum and I would have to do all the validation myself or write glue code. If I were to write usual handlers I would get to type safety with it and would be sure that the header I have defined is actually in place as this function is called. And as a library author wanting to give full and usual axum capabilities this is just a huge blocker as now there is no way to have that promise.

Is this by design?

PS: Didn't open an issue as from me its a general question

**Comment by davidpdrsn** [comment, marked as answer]
See https://docs.rs/axum/latest/axum/extract/index.html#applying-multiple-extractors. You need to put the `Request` extractor last.

**Reply by somehowchris** [reply]
Thanks 😊 

The docs are really nice and extensive but as a new one to axum I find it a bit hard to find the right pieces. Is there something in the docs.rs search where not only parameters, names and return types are searched but also all the documentation?

**Reply by davidpdrsn** [reply]
I don't think so.

Generally axum's docs is structured such that each top level module gives an overall idea of how to use that subset, since many of the types inside are used in similar ways.

So I'd recommend reading the docs for each top level module to get a good understand of how things fit together.

You're also welcome to make PRs with suggestions for improvements 😊

**Reply by somehowchris** [reply]
Don’t get me wrong, the docs are awesome but the docs.rs search for this kind of extensive documentation doesn’t really seem UX friendly as its more like a types search rather than a doc search

**Reply by davidpdrsn** [reply]
You should file an issue with rustdoc 😊
